R/course.R

```diff
@@ -259,6 +264,81 @@ tidy_download <- function(url, destdir = getwd()) {
   invisible(full_path)
 }
 
+download_url <- function(url,
+                         destfile,
+                         handle = curl::new_handle(),
+                         n_tries = 3,
+                         retry_connecttimeout = 40L) {
+  handle_options <- list(noprogress = FALSE, progressfunction = progress_fun)
+  curl::handle_setopt(handle, .list = handle_options)
+
+  we_should_retry <- function(i, n_tries, status) {
+    if (i >= n_tries) {
+      FALSE
+    } else if (inherits(status, "error")) {
+      # TODO: find a way to detect a (connect) timeout more specifically?
+      # https://github.com/jeroen/curl/issues/154
+      # https://ec.haxx.se/usingcurl/usingcurl-timeouts
+      # "Failing to connect within the given time will cause curl to exit with a
+      # timeout exit code (28)."
+      # (however, note that all timeouts lead to this same exit code)
+      # https://ec.haxx.se/usingcurl/usingcurl-returns
+      # "28. Operation timeout. The specified time-out period was reached
+      # according to the conditions. curl offers several timeouts, and this exit
+      # code tells one of those timeout limits were reached."
+      # https://github.com/curl/curl/blob/272282a05416e42d2cc4a847a31fd457bc6cc827/lib/strerror.c#L143-L144
+      # "Timeout was reached" <-- actual message we could potentially match
+      TRUE
+    } else {
+      FALSE
+    }
+  }
+
+  status <- try_download(url, destfile, handle = handle)
+  if (inherits(status, "error") && interactive()) {
+    ui_oops(status$message)
+    if (ui_nope("
+      Download failed :(
+      See above for everything we know about why it failed.
+      Shall we try a couple more times, with a longer timeout?
+    ")) {
+      n_tries <- 1
+    }
+  }
+
+  i <- 1
+  # invariant: we have made i download attempts
+  while (we_should_retry(i, n_tries, status)) {
+    if (i == 1) {
+      curl::handle_setopt(
+        handle,
+        .list = c(connecttimeout = retry_connecttimeout))
```

Comment:
I have no idea how I can test that this is actually having the desired effect.

Reply:
You can try to connect to a dead host a look at `system.time()` afterwards maybe?

Reply:
Do you have any ideas of a specific "dead host"? Or method for finding one? If I simply disconnect from the internet, I get an immediate error about not being able to resolve the host name.

Reply:
It's a bit tricky indeed because you need a host that exists but does not respond. You could try to connect to a port that is not used. For example:

```r
h <- new_handle(connecttimeout=7, verbose=T)
system.time(curl_fetch_memory('http://cran.r-project.org:999', handle = h))
```

But this is also not waterproof because some servers might have a firewall that explicitly rejects the connection so it fails immediately. But the example above seems to work for me.

Reply:
Thanks, that example is pretty compelling. I mostly wanted to confirm that the name `connecttimeout` is correct and that I'm using `curl::handle_setopt()` correctly. So I think we're good on those fronts.

``` r
library(curl)

h <- new_handle(connecttimeout = 3, verbose = TRUE)
system.time(curl_fetch_memory('http://cran.r-project.org:999', handle = h))
#> Error in curl_fetch_memory("http://cran.r-project.org:999", handle = h): Timeout was reached: [cran.r-project.org:999] Connection timed out after 3003 milliseconds
#> Timing stopped at: 0.017 0.007 3.005

handle_setopt(h, connecttimeout = 6)
system.time(curl_fetch_memory('http://cran.r-project.org:999', handle = h))
#> Error in curl_fetch_memory("http://cran.r-project.org:999", handle = h): Timeout was reached: [cran.r-project.org:999] Connection timed out after 6003 milliseconds
#> Timing stopped at: 0.001 0 6.003

handle_setopt(h, .list = list(connecttimeout = 5))
system.time(curl_fetch_memory('http://cran.r-project.org:999', handle = h))
#> Error in curl_fetch_memory("http://cran.r-project.org:999", handle = h): Timeout was reached: [cran.r-project.org:999] Connection timed out after 5004 milliseconds
#> Timing stopped at: 0.001 0.001 5.005
```

<sup>Created on 2020-03-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>